src/ServiceControl.Monitoring/QueueLength/QueueLengthCalculator.cs

```diff
@@ -3,45 +3,81 @@
     using System;
     using System.Collections.Concurrent;
     using System.Collections.Generic;
+    using System.Linq;
 
     class QueueLengthCalculator : IQueueLengthCalculator
     {
-        public void UpdateReceivedSequence(string key, long value, string queue)
+        public bool UpdateReceivedSequence(string endpointName, string key, double value, string queue)
         {
-            receivedSequences.AddOrUpdate(key, _ => value, (_, prev) => Math.Max(prev, value));
-            sessionKeyToQueueMapping.GetOrAdd(key, _ => queue);
+            var virtualQueue = new VirtualQueueId
+            {
+                EndpointName = endpointName,
+                QueueName = queue,
+                SessionKey = key
+            };
+
+            receivedSequences.AddOrUpdate(virtualQueue, _ => value, (_, previousValue) => Math.Max(previousValue, value));
+
+            return sentSequences.ContainsKey(key);
         }
 
-        public void UpdateSentSequence(string key, long value)
+        public List<VirtualQueueId> UpdateSentSequence(string key, double value)
         {
-            sentSequences.AddOrUpdate(key, _ => value, (_, prev) => Math.Max(prev, value));
+            sentSequences.AddOrUpdate(key, _ => value, (_, previousValue) => Math.Max(previousValue, value));
+
+            return receivedSequences.Keys.Where(k => k.SessionKey == key).ToList();
         }
 
-        public Dictionary<string, long> GetQueueLengths()
+        public Dictionary<VirtualQueueId, double> GetQueueLengths()
         {
-            var queueLengths = new Dictionary<string, long>();
+            var queueLengths = new Dictionary<VirtualQueueId, double>();
 
-            foreach (var kvp in sessionKeyToQueueMapping)
+            foreach (var virtualQueueId in receivedSequences.Keys)
```

Comment:
After going through the implementation of a concurrent dictionary, I'd say, that we should use `foreach` over the dictionary rather than accessing it's keys to get the query. Creating an enumerator over concurrent dictionary uses just volatile reads, without issuing any locks. As far as I see it, it'd be suitable for this situation. Additionally, we'd receive both the key and the value in one read.